packages/@lwc/engine-server/src/context.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import {
+    createContextProviderWithRegister,
+    getAssociatedVMIfPresent,
+    LightningElement,
+    WireAdapterConstructor,
+    WireContextSubscriptionPayload,
+    WireContextSubscriptionCallback,
+} from '@lwc/engine-core';
+import { isUndefined } from '@lwc/shared';
+import {
+    HostElement,
+    HostNodeType,
+    HostTypeKey,
+    HostParentKey,
+    HostParentNode,
+    HostHostKey,
+    HostContextProvidersKey,
+} from './types';
+
+export function createContextProvider(adapter: WireAdapterConstructor) {
+    return createContextProviderWithRegister(adapter, registerContextProvider);
+}
+
+function registerContextProvider(
+    elm: HostElement | LightningElement,
+    adapterContextToken: string,
+    onContextSubscription: WireContextSubscriptionCallback
+) {
+    const vm = getAssociatedVMIfPresent(elm);
+    if (!isUndefined(vm)) {
+        elm = vm.elm;
+    }
+
+    const contextProviders = (elm as HostElement)[HostContextProvidersKey];
+    if (!contextProviders) {
+        throw new Error('Unable to register context provider with provided `elm`.');
```

Comment:
This error is not tested anywhere. Is it possible to hit this condition?

Reply:
This would only happen if the author tried to install a context provider on a non-Element. I can add a test case for this.

Reply:
Added test: `context-installation-on-non-element`